pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,30 +91,147 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
+// SaveAlertInstances saves all the provided alert instances to the store.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	if !st.FeatureToggles.IsEnabled(featuremgmt.FlagAlertingBigTransactions) {
+		// This mimics the replace code-path by calling SaveAlertInstance in a loop, with a transaction per call.
+		for _, c := range cmd {
+			err := st.SaveAlertInstance(ctx, c)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	} else {
+		// The function starts a single transaction and batches writes into
+		// statements with `maxRows` instances per statements. This makes for a
+		// fairly efficient transaction without creating statements that are too long
+		// for some databases to process. For example, SQLite has a limit of 999
+		// variables per write.
+		err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
```

Comment:
Do we really need a transaction for this? If we do, perhaps we could write each batch as an transaction instead. Long running transaction blocks the database pool

Reply:
cc @grobinson-grafana we went back and forth on this one in the original PR. The existing code executes a separate transaction per alert instance, which is the basic inefficiency we were looking to get around. 

There are correctness issues with the current code, and with any code that doesn't do all the writes in a single transaction - you can end up getting partial alert state rather than rolling the whole state back on a database error. But that's a problem we already live with, and bigger chunks per transaction make that better, not worse.

Doing it in batches is up to 20x fewer transactions - I'm willing to bet that *most* calls to this function will be one to three transactions, heavily weighted towards "one", so a tighter limit on how long transactions can get seems reasonable to me.

Reply:
Doesn't `withDbSession` also keep the connection until the session is closed? I'm not sure I understand the difference between having a long running session and a long running transaction when it comes to returning connections to the connection pool?

Reply:
We have 5 connections per instance in Hosted Grafana, and 1 of those connections might be utilized writing the alert state to the database; however instead of doing that we want to distribute the operation over a larger number of short-lived sessions?

Reply:
Many short-lived sessions is what we were doing already - creating a session to write each individual alert instance. It's not the case that we only ever use a single connection, because multiple alerting rules may be writing to the database concurrently. It's possible that alert instance writes could consume all 5 connections in some cases.

It's not clear yet which granularity of transactions would be best for our usage, and we don't have automated testing available yet to evaluate the tradeoff in a realistic environment, unfortunately.